### DISCUSSION TITLE
Adding context processors for templates

### DISCUSSION THREAD

**Discussion by jtraub** [litestar-org/litestar#653] [Ideas]
Sometimes all templates share same variables (for example, username to display profile menu in the top right corner or number of items in cart). Straightforward approach would be calculating values and passing them to `Template` in `context` kwarg in each route handler.

Better approach would be adding functions that take `Request` as an argument and return dict that should be merged into template context. These functions should be execute by Starlite before template rendering.

Something like this
```python
def user_info(request: Request):
      return {"username": request.user.name, "role": request.user.role}

template_config.engine.register_template_context_processor(user_info)

# {{ username }} and {{ role }} are now available in every template
```

Other frameworks offer similar concept:
* [View sharing](https://docs.masoniteproject.com/the-basics/views#view-sharing) in Masonite
* [Context processors](https://docs.djangoproject.com/en/4.1/topics/templates/#context-processors) in Django
* [Context processors](https://flask.palletsprojects.com/en/2.2.x/templating/#context-processors) in Flask

We already have [Template Callables](https://starlite-api.github.io/starlite/usage/16-templating/2-template-functions/#registering-template-callables) in Starlite which can be used to resolve the problem. However, there are two downsides:
* Callable will be executed each time it is accessed in a template. So the user will have to apply some kind of memoization if a variable they want to share requires expensive computations and is being used multiple times in the template.
* Callable always returns a single value so if computation derives multiple values they should be returned as a dict making templates code more complicated in case you need .

I am hesitant about context processors.

**Comment by Goldziher** [comment]
why are you hesitent?

**Reply by jtraub** [reply]
I like to keep API surface small. Template callables kinda solve the problem but require more skill to not become a footgun.

Basically my question boils down to picking between keeping small API or improving user experience.

**Reply by provinzkraut** [reply]
@jtraub I think this is a worthwhile feature, since it's very commonly used.
